rasa/core/training/generator.py

```diff
@@ -82,24 +83,34 @@ def from_events(
             tracker.update(e)
         return tracker
 
-    def past_states(self, domain: Domain) -> deque:
-        """Return the states of the tracker based on the logged events."""
-
+    def past_frozen_states(self, domain: Domain) -> Deque[FrozenState]:
         # we need to make sure this is the same domain, otherwise things will
         # go south. but really, the same tracker shouldn't be used across
         # domains
         assert domain == self.domain
 
         # if don't have it cached, we use the domain to calculate the states
         # from the events
-        if self._states is None:
-            self._states = super().past_states(domain)
+        if self._frozen_states is None:
+            states = super().past_states(domain)
+            self._frozen_states = deque(self.freeze_current_state(s) for s in states)
+
+        return self._frozen_states
+
+    @staticmethod
+    def _unfreeze_states(frozen_states: Deque[FrozenState]) -> List[State]:
+        return [
+            {key: dict(value) for key, value in dict(frozen_state).items()}
+            for frozen_state in frozen_states
+        ]
 
-        return self._states
+    def past_states(self, domain: Domain) -> List[State]:
+        frozen_states = self.past_frozen_states(domain)
+        return self._unfreeze_states(frozen_states)
 
     def clear_states(self) -> None:
         """Reset the states."""
-        self._states = None
+        self._frozen_states = None
```

Comment:
to make sure, that frozen states are not used as actual states, since they are actually different thing `FrozenState` vs `State`. why you don't like the rename?

Reply:
renamed to `_states_for_hashing`